### PR TITLE
gh-91181: restore support for bytes on sys.path in FileFinder

### LINKED ISSUE #91181: bytes do not work on sys.path
BPO | [47025](https://bugs.python.org/issue47025)
--- | :---
Nosy | @brettcannon, @jaraco, @ericvsmith, @eryksun, @graingert
PRs | <li>python/cpython#31897</li><li>python/cpython#31934</li>
Files | <li>[demo.py](https://bugs.python.org/file50678/demo.py "Uploaded as text/plain at 2022-03-15.10:32:43 by @graingert"): reproducer of the failure</li><li>[zipfile_demo.py](https://bugs.python.org/file50679/zipfile_demo.py "Uploaded as text/plain at 2022-03-15.11:24:43 by @graingert")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-03-15.10:32:43.256>
labels = ['library', '3.9', '3.10', '3.11']
title = 'bytes do not work on sys.path'
updated_at = <Date 2022-03-17.19:22:31.783>
user = 'https://github.com/graingert'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-17.19:22:31.783>
actor = 'brett.cannon'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2022-03-15.10:32:43.256>
creator = 'graingert'
dependencies = []
files = ['50678', '50679']
hgrepos = []
issue_num = 47025
keywords = ['patch']
message_count = 10.0
messages = ['415233', '415234', '415235', '415236', '415237', '415238', '415239', '415323', '415340', '415443']
nosy_count = 5.0
nosy_names = ['brett.cannon', 'jaraco', 'eric.smith', 'eryksun', 'graingert']
pr_nums = ['31897', '31934']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue47025'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-91181

### CODE REVIEW

File: Lib/importlib/_bootstrap_external.py [hunk 1]
```diff
@@ -1457,7 +1468,9 @@ def _get_spec(cls, fullname, path, target=None):
         #  the list of paths that will become its __path__
         namespace_path = []
         for entry in path:
-            if not isinstance(entry, (str, bytes)):
+            if isinstance(entry, bytes):
+                entry = _fsdecode(entry)
```
Comment (graingert) [inline Lib/importlib/_bootstrap_external.py]: perhaps this should ignore errors?

```suggestion
                try:
                    entry = _fsdecode(entry)
                except UnicodeDecodeError:
                    continue
```

File: Lib/test/test_importlib/import_/test_path.py [hunk 2]
```diff
@@ -116,6 +116,31 @@ def test_None_on_sys_path(self):
...
+                              self.machinery.FileFinder.path_hook(
+                                  *self.importlib._bootstrap_external._get_supported_file_loaders())]
+            missing = object()
+            email = sys.modules.pop('email', missing)
+            try:
+                with util.import_state(meta_path=sys.meta_path[:],
+                                       path=new_path,
+                                       path_importer_cache=new_path_importer_cache,
+                                       path_hooks=new_path_hooks):
+                    module = self.importlib.import_module('email')
```
Comment (graingert) [inline Lib/test/test_importlib/import_/test_path.py:137]: this results in a `BytesWarning: Comparison between bytes and string` when bytes are added to the `sys.path_importer_cache` this is an existing problem for zipimport paths https://bugs.python.org/issue47026

File: Misc/NEWS.d/next/Library/2022-03-15-11-00-55.bpo-47025.oS8dt6.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+support bytes on sys.path in FileFinder
```
Comment (brettcannon) [inline Misc/NEWS.d/next/Library/2022-03-15-11-00-55.bpo-47025.oS8dt6.rst]:
```suggestion
support bytes on sys.path in FileFinder. Patch by Thomas Grainger.
```

**Review by brettcannon — APPROVED** [review]

File: Lib/test/test_importlib/import_/test_path.py (hunk 2 above)
Comment (brettcannon) [reply Lib/test/test_importlib/import_/test_path.py:137]: Looks like that's causing CI to fail as `BytesWarning` is considered an error. Looks like I need to go to python-dev and ask what people want to do.

Comment (brettcannon) [reply Lib/test/test_importlib/import_/test_path.py:137]: https://discuss.python.org/t/drop-supporting-bytes-on-sys-path/17225

I'm now suggesting dropping bytes support in `sys.path`. We will see how that conversation goes.

**Comment by graingert** [comment]
Closing in favor of https://github.com/python/cpython/pull/31934

**Pull request closed without merging**